File: lib/tooltip/tooltip.dart
import "package:angular2/angular2.dart";
import "package:ng2_strap/position.dart" show positionService;
import 'dart:async';

class TooltipOptions {
  String placement;

  String popupClass;

  bool animation;

  bool isOpen;

  var content;

  TooltipOptions({
    this.placement,
    this.popupClass,
    this.animation,
    this.isOpen,
    this.content
  });
}

@Component (selector: "n2s-tooltip-container",
    template: '''
    <div class="tooltip" role="tooltip"
     [ngStyle]="{top: top, left: left, display: display}"
     [ngClass]="classMap" >
      <div class="tooltip-arrow"></div>
      <div class="tooltip-inner">
        {{content}}
      </div>
    </div>''',
    directives: const [NgClass, NgStyle],
    encapsulation: ViewEncapsulation.None)
class TooltipContainer {
  ElementRef element;

  Map<String, dynamic> classMap;

  dynamic positionMap;

  String top;

  String left;

  String display;

  String content;

  String placement = "top";

  bool appendToBody = false;

  bool isOpen;

  String popupClass;

  bool animation;

  TooltipContainer(this.element, TooltipOptions options) {
    classMap = { "in" : false};
    placement = options.placement;
    popupClass = options.popupClass;
    animation = options.animation;
    isOpen = options.isOpen;
    content = options.content;
    classMap[placement] = true;
  }

  position(ElementRef hostEl) {
    display = "block";
    top = "0px";
    left = "0px";
    var p = positionService.positionElements(
        hostEl.nativeElement,
        element.nativeElement.children[0],
        placement,
        appendToBody);
    top = p.top.toString() + "px";
    left = p.left.toString() + "px";
    classMap [ "in" ] = true;
  }
}

@Directive (selector: "[tooltip]",
    inputs: const [
      "content:tooltip",
      "placement:tooltip-placement",
      "appendToBody",
      "isOpen: tooltip-is-open",
      "enable: tooltip-enable"
    ],
    host: const {
      "(mouseenter)" : "show(\$event)",
      "(mouseleave)" : "hide(\$event)",
      "(focusin)" : "show(\$event)",
      "(focusout)" : "hide(\$event)"
    })
class Tooltip {
  ElementRef element;

  DynamicComponentLoader loader;

  bool visible = false;

  String content;

  String placement = "top";

  // todo:
  bool appendToBody;

  bool isOpen;

  bool enable;

  Future<ComponentRef> tooltip;

  Tooltip(this.element, this.loader);

  // todo: filter triggers
  show(event) {
    if (visible) {
      return;
    }
    visible = true;
    var options = new TooltipOptions (content: content, placement: placement);
    var binding = Injector.resolve([bind(TooltipOptions).toValue(options)]);
    tooltip = loader.loadNextToLocation(TooltipContainer, element, binding)
        .then((ComponentRef componentRef) {
      return new Future.delayed(const Duration(milliseconds: 1), () {
        (componentRef.instance as TooltipContainer).position(element);
        return componentRef;
      });
    });
  }

  hide(event) {
    if (!visible) {
      return;
    }
    visible = false;
    tooltip.then((ComponentRef componentRef) {
      componentRef.dispose();
      return componentRef;
    });
  }
}

const List<dynamic> TOOLTIP_DIRECTIVES = const [Tooltip, TooltipContainer];

File: components/tooltip/tooltip.ts
import {
  Directive, Component,
  OnInit, Input, HostListener,
  ElementRef, EventEmitter,
  DynamicComponentLoader, ComponentRef
} from 'angular2/core';
import { NgClass, NgStyle } from 'angular2/common';
import { bind, Injectable, forwardRef, ResolvedBinding, Injector } from 'angular2/core';

import { positionService } from '../position';
import { IAttribute } from '../common';

class TooltipOptions {
  public placement:string;
  public popupClass:string;
  public animation:boolean;
  public isOpen:boolean;

  constructor(options:Object) {
    Object.assign(this, options);
  }
}

@Component({
  selector: 'tooltip-container',
  directives: [NgClass, NgStyle],
  template: `
    <div class="tooltip" role="tooltip"
     [ngStyle]="{top: top, left: left, display: display}"
     [ngClass]="classMap" >
      <div class="tooltip-arrow"></div>
      <div class="tooltip-inner">
        {{content}}
      </div>
    </div>`
})
class TooltipContainer {
  private classMap:any;
  private positionMap:any;
  private top:string;
  private left:string;
  private display:string;
  private content:string;
  private placement:string;
  private appendToBody:boolean;

  private isOpen:boolean;

  constructor(public element:ElementRef, options:TooltipOptions) {
    Object.assign(this, options);
    this.classMap = {'in': false};
    this.classMap[options.placement] = true;
  }

  public position(hostEl:ElementRef) {
    this.display = 'block';
    this.top = '0px';
    this.left = '0px';
    let p = positionService
      .positionElements(hostEl.nativeElement,
        this.element.nativeElement.children[0],
        this.placement, this.appendToBody);
    this.top = p.top + 'px';
    this.left = p.left + 'px';
    this.classMap['in'] = true;
  }
}

@Directive({selector: '[tooltip]'})
export class Tooltip implements OnInit {
  @Input('tooltip') private content:string;
  @Input('tooltip-placement') private placement:string = 'top';
  @Input('tooltip-isOpen') private isOpen:boolean;
  @Input('tooltip-enable') private enable:boolean;
  @Input() private appendToBody:boolean;

  private visible:boolean = false;
  private tooltip:Promise<ComponentRef>;

  constructor(public element:ElementRef,
              public loader:DynamicComponentLoader) {
  }

  ngOnInit() {
  }

  // todo: filter triggers
  // params: event, target
  @HostListener('focusin', ['$event', '$target'])
  @HostListener('mouseenter', ['$event', '$target'])
  show() {
    if (this.visible) {
      return;
    }
    this.visible = true;

    let options = new TooltipOptions({
      content: this.content,
      placement: this.placement
    });

    let binding = Injector.resolve([
      bind(TooltipOptions).toValue(options)
    ]);

    this.tooltip = this.loader
      .loadNextToLocation(TooltipContainer, this.element, binding)
      .then((componentRef:ComponentRef) => {
        componentRef.instance.position(this.element);
        return componentRef;
      });
  }

  // params event, target
  @HostListener('focusout', ['$event', '$target'])
  @HostListener('mouseleave', ['$event', '$target'])
  hide() {
    if (!this.visible) {
      return;
    }
    this.visible = false;
    this.tooltip.then((componentRef:ComponentRef) => {
      componentRef.dispose();
      return componentRef;
    });
  }
}

export const TOOLTIP_DIRECTIVES:Array<any> = [Tooltip, TooltipContainer];
/**
 * @deprecated
 * @type {Tooltip|TooltipContainer[]}
 */
export const tooltip:Array<any> = [Tooltip, TooltipContainer];
